File: testers/sqlserver/scala/src/scala/testdb/ForeignKeyTest.scala
package testdb

import org.junit.Assert._
import org.junit.Test
import testdb.customers._
import testdb.order_items._
import testdb.orders._
import testdb.products._
import testdb.userdefined.Email

import scala.util.Random

/** Tests for foreign key relationships between tables. Tests orders -> customers and order_items -> orders/products relationships.
  */
class ForeignKeyTest {
  private val testInsert = TestInsert(Random(42))
  private val customersRepo = CustomersRepoImpl()
  private val productsRepo = ProductsRepoImpl()
  private val ordersRepo = OrdersRepoImpl()
  private val orderItemsRepo = OrderItemsRepoImpl()

  @Test
  def testOrderReferencesCustomer(): Unit = withConnection { c =>
    given java.sql.Connection = c

    val customer = testInsert.Customers(email = Email("order-ref@example.com"))
    val order = testInsert.Orders(customerId = customer.customerId, totalAmount = BigDecimal("99.99"))

    assertNotNull(order)
    assertEquals(customer.customerId, order.customerId)

    val foundCustomer = customersRepo.selectById(order.customerId)
    assertTrue(foundCustomer.isDefined)
    assertEquals(customer.name, foundCustomer.get.name)
  }

  @Test
  def testOrderItemsReferencesOrderAndProduct(): Unit = withConnection { c =>
    given java.sql.Connection = c

    val customer = testInsert.Customers(email = Email("item-refs@example.com"))
    val product = testInsert.Products()
    val order = testInsert.Orders(customerId = customer.customerId)

    val orderItem = testInsert.OrderItems(
      orderId = order.orderId,
      productId = product.productId,
      quantity = 5,
      price = BigDecimal("19.99")
    )

    assertNotNull(orderItem)
    assertEquals(order.orderId, orderItem.orderId)
    assertEquals(product.productId, orderItem.productId)

    val foundOrder = ordersRepo.selectById(orderItem.orderId)
    assertTrue(foundOrder.isDefined)

    val foundProduct = productsRepo.selectById(orderItem.productId)
    assertTrue(foundProduct.isDefined)
  }

  @Test
  def testMultipleOrdersForCustomer(): Unit = withConnection { c =>
    given java.sql.Connection = c

    val customer = testInsert.Customers(email = Email("multi-orders@example.com"))

    val order1 = testInsert.Orders(customerId = customer.customerId, totalAmount = BigDecimal("100.00"))
    val order2 = testInsert.Orders(customerId = customer.customerId, totalAmount = BigDecimal("200.00"))
    val order3 = testInsert.Orders(customerId = customer.customerId, totalAmount = BigDecimal("300.00"))

    assertEquals(customer.customerId, order1.customerId)
    assertEquals(customer.customerId, order2.customerId)
    assertEquals(customer.customerId, order3.customerId)

    assertNotEquals(order1.orderId, order2.orderId)
    assertNotEquals(order2.orderId, order3.orderId)
  }

  @Test
  def testMultipleItemsForOrder(): Unit = withConnection { c =>
    given java.sql.Connection = c

    val customer = testInsert.Customers(email = Email("multi-items@example.com"))
    val product1 = testInsert.Products()
    val product2 = testInsert.Products()
    val order = testInsert.Orders(customerId = customer.customerId)

    val item1 = testInsert.OrderItems(
      orderId = order.orderId,
      productId = product1.productId,
      quantity = 2,
      price = BigDecimal("10.00")
    )
    val item2 = testInsert.OrderItems(
      orderId = order.orderId,
      productId = product2.productId,
      quantity = 3,
      price = BigDecimal("15.00")
    )

    assertEquals(order.orderId, item1.orderId)
    assertEquals(order.orderId, item2.orderId)
    assertEquals(product1.productId, item1.productId)
    assertEquals(product2.productId, item2.productId)
  }

  @Test
  def testCascadingForeignKeys(): Unit = withConnection { c =>
    given java.sql.Connection = c

    val customer = testInsert.Customers(email = Email("cascade-fk@example.com"))
    val product = testInsert.Products()
    val order = testInsert.Orders(customerId = customer.customerId)
    val orderItem = testInsert.OrderItems(
      orderId = order.orderId,
      productId = product.productId,
      price = BigDecimal("25.00")
    )

    val foundItem = orderItemsRepo.selectById(orderItem.orderItemId)
    assertTrue(foundItem.isDefined)

    val foundOrder = ordersRepo.selectById(foundItem.get.orderId)
    assertTrue(foundOrder.isDefined)

    val foundCustomer = customersRepo.selectById(foundOrder.get.customerId)
    assertTrue(foundCustomer.isDefined)

    assertEquals(customer.customerId, foundCustomer.get.customerId)
  }
}

File: testers/db2/java/src/java/testdb/ForeignKeyTest.java
package testdb;

import static org.junit.Assert.*;

import java.math.BigDecimal;
import java.util.Optional;
import java.util.Random;
import org.junit.Test;
import testdb.customers.*;
import testdb.order_items.*;
import testdb.orders.*;

/**
 * Tests for foreign key relationships between tables. Tests orders -> customers and order_items ->
 * orders relationships.
 */
public class ForeignKeyTest {
  private final TestInsert testInsert = new TestInsert(new Random(744282577));
  private final CustomersRepoImpl customersRepo = new CustomersRepoImpl();
  private final OrdersRepoImpl ordersRepo = new OrdersRepoImpl();
  private final OrderItemsRepoImpl orderItemsRepo = new OrderItemsRepoImpl();

  @Test
  public void testOrderReferencesCustomer() {
    Db2TestHelper.run(
        c -> {
          var customer = testInsert.Customers().insert(c);
          var order =
              testInsert
                  .Orders(customer.customerId())
                  .with(r -> r.withTotalAmount(Optional.of(new BigDecimal("99.99"))))
                  .insert(c);

          assertNotNull(order);
          assertEquals(customer.customerId(), order.customerId());

          var foundCustomer = customersRepo.selectById(order.customerId(), c);
          assertTrue(foundCustomer.isPresent());
          assertEquals(customer.name(), foundCustomer.get().name());
        });
  }

  @Test
  public void testOrderItemsReferencesOrder() {
    Db2TestHelper.run(
        c -> {
          var customer = testInsert.Customers().insert(c);
          var order = testInsert.Orders(customer.customerId()).insert(c);

          var orderItem =
              testInsert
                  .OrderItems(order.orderId())
                  .with(r -> r.withQuantity(5).withUnitPrice(new BigDecimal("19.99")))
                  .insert(c);

          assertNotNull(orderItem);
          assertEquals(order.orderId(), orderItem.orderId());

          var foundOrder = ordersRepo.selectById(orderItem.orderId(), c);
          assertTrue(foundOrder.isPresent());
        });
  }

  @Test
  public void testMultipleOrdersForCustomer() {
    Db2TestHelper.run(
        c -> {
          var customer = testInsert.Customers().insert(c);

          var order1 =
              testInsert
                  .Orders(customer.customerId())
                  .with(r -> r.withTotalAmount(Optional.of(new BigDecimal("100.00"))))
                  .insert(c);
          var order2 =
              testInsert
                  .Orders(customer.customerId())
                  .with(r -> r.withTotalAmount(Optional.of(new BigDecimal("200.00"))))
                  .insert(c);
          var order3 =
              testInsert
                  .Orders(customer.customerId())
                  .with(r -> r.withTotalAmount(Optional.of(new BigDecimal("300.00"))))
                  .insert(c);

          assertEquals(customer.customerId(), order1.customerId());
          assertEquals(customer.customerId(), order2.customerId());
          assertEquals(customer.customerId(), order3.customerId());

          assertNotEquals(order1.orderId(), order2.orderId());
          assertNotEquals(order2.orderId(), order3.orderId());
        });
  }

  @Test
  public void testMultipleItemsForOrder() {
    Db2TestHelper.run(
        c -> {
          var customer = testInsert.Customers().insert(c);
          var order = testInsert.Orders(customer.customerId()).insert(c);

          var item1 = testInsert.OrderItems(order.orderId()).with(r -> r.withQuantity(2)).insert(c);
          var item2 = testInsert.OrderItems(order.orderId()).with(r -> r.withQuantity(3)).insert(c);

          assertEquals(order.orderId(), item1.orderId());
          assertEquals(order.orderId(), item2.orderId());
        });
  }

  @Test
  public void testCascadingForeignKeys() {
    Db2TestHelper.run(
        c -> {
          var customer = testInsert.Customers().insert(c);
          var order = testInsert.Orders(customer.customerId()).insert(c);
          var orderItem = testInsert.OrderItems(order.orderId()).insert(c);

          var foundItem = orderItemsRepo.selectById(orderItem.id(), c);
          assertTrue(foundItem.isPresent());

          var foundOrder = ordersRepo.selectById(foundItem.get().orderId(), c);
          assertTrue(foundOrder.isPresent());

          var foundCustomer = customersRepo.selectById(foundOrder.get().customerId(), c);
          assertTrue(foundCustomer.isPresent());

          assertEquals(customer.customerId(), foundCustomer.get().customerId());
        });
  }
}
